### ISSUE TITLE
Ability to record sub stream along with main stream

### ISSUE THREAD

**Issue by darkalfx** [blakeblackshear/frigate#2082]
**Describe what you are trying to accomplish and why in non technical terms**

I have cameras with very high resolutions, I want to be able to do continuous recording of the lower resolution input stream but whenever a detection occurs, I want the associated clips to be taken from the high resolution input stream, not from the lower resolution input stream.

**Describe the solution you'd like**

A new camera role called something like "Clips" would allow users to choose which input stream to use for generating clips instead of assuming that this should come from the continous recording role.

**Describe alternatives you've considered**

I could duplicate the camera in Frigate to achieve this, however this would also mean running detection twice for the same stream which is highly undesirable.

I could also do continuous recording of the high resolution input stream but that would take very large amounts of disk space, not ideal.

**Comment by finnaGIT** [comment]
Below is a proposed config to implement continuous recording of sub-stream and recording only `active_objects` in the main stream. Of course while I see this as easy to configure from the user's side, I have no idea if this would work behind the scenes. Just for consideration:

```
go2rtc:
  streams:
    camera_one_main: rtsp://192.168.0.2:1234 # camera one main stream
    camera_one_sub: rtsp://192.168.0.2:5678 # camera one sub stream

cameras:
  camera_one:
    enabled: true
    ffmpeg:
      inputs:
        - path: rtsp://localhost:8554/camera_one_main
          roles:
            - record
            - audio
        - path: rtsp://localhost:8554/camera_one_sub
          roles:
            - detect
            - record
            - audio
    record:
      enabled: true
      retain:
        mode: all
        days: -1 # '-1' to record continuously, or until storage is consumed & delete oldest 2 hours with 1 hr remaining, or state a number of days as is currently implemented
        stream: camera_one_sub
      alerts:
        mode: active_objects
        days: -1
        stream: camera_one_main
```

**Comment by partytimeexcellent** [comment]
My issue is the ability to stream high-res from my house with only 11Mb/s upload speed, which can buffer clips for 20-30 seconds before playback. Basically as it sits now, I can't use my 4k cameras to their full potential.

My idea for this implementation would be the ability to have a high-res recording accessible on the LAN, and a secondary low-res recording for remote viewing. Possibly some logic to detect if I'm on the LAN or remote when presenting the stream.

A secondary option would be on-the-fly transcoding from high-res to low-res when remote viewing, possibly with logic to detect when streaming is too slow. Seems like this is how some other NVR software handles it.

**Comment by mdv0** [comment]
Has anyone figured out a way to always record the sub stream and use the main stream only for events? Or do the devs have any plans to implement this?

**Comment by NickM-27** [comment]
Yes, that's why it's pinned

**Comment by stefa168** [comment]
I've been trying Frigate for some time now, alongside my main Blue Iris instance. This is the main feature I'm missing in Frigate.

In my opinion, it would be preferable to have to tag the two input streams as "main" and "sub" and then just having Frigate use them as such

**Comment by dominikgrygiel** [comment]
> Has anyone figured out a way to always record the sub stream and use the main stream only for events? Or do the devs have any plans to implement this?

I was able to achieve it by duplicating the camera:

```yaml
cameras:
  camera_one_continuous:
    enabled: true
    ffmpeg:
      inputs:
        - path: 
            rtsp://secondary_stream_url
          roles:
            - record
    detect:
      enabled: false
    record:
      enabled: true
      retain:
        days: 7
        mode: all
  camera_one:
    enabled: true
    ffmpeg:
      inputs:
        - path: 
            rtsp://main_stream_url
          roles:
            - record
        - path: 
            rtsp://secondary_stream_url
          roles:
            - detect
    detect:
      enabled: true
    record:
      enabled: true
      retain:
        days: 30
        mode: motion
```

**Comment by kabaga** [comment]
> > Has anyone figured out a way to always record the sub stream and use the main stream only for events? Or do the devs have any plans to implement this?
> 
> I was able to achieve it by duplicating the camera:
> 
> cameras:
>   camera_one_continuous:
>     enabled: true
>     ffmpeg:
>       inputs:
>         - path: 
>             rtsp://secondary_stream_url
>           roles:
>             - record
>     detect:
>       enabled: false
>     record:
>       enabled: true
>       retain:
>         days: 7
>         mode: all
>   camera_one:
>     enabled: true
>     ffmpeg:
>       inputs:
>         - path: 
>             rtsp://main_stream_url
>           roles:
>             - record
>         - path: 
>             rtsp://secondary_stream_url
>           roles:
>             - detect
>     detect:
>       enabled: true
>     record:
>       enabled: true
>       retain:
>         days: 30
>         mode: motion

I tried to replicate this on version 0.16 and it appears that the main stream is still continuing to record even though it is set to motion. When I go to the Live / History, and scrub the timeline, i am able to play the video on the main stream. The only difference I could tell between the continues sub stream and continues (supposed to be just motion) main stream, the main stream has a lot of motion yellow horizontal lines in its timeline.
Here is my settings:
```
cameras:
  livingroom_sub_continues:
    enabled: true
    ui:
      dashboard: false
    ffmpeg:
      inputs:
        - path: rtsp://localhost:8554/livingroom_sub
          input_args: preset-rtsp-restream
          roles:
            - record
            - audio
    record:
      enabled: true
      retain:
        days: 2
        mode: all
    detect:
      enabled: true
      width: 640
      height: 480
      fps: 4
    objects:
      track:
        - person
        - dog
      filters:
        person:
          min_score: 0.4
          threshold: 0.7
          min_area: 700
    motion:
      mask: 0.493,0.019,0.594,0.017,0.594,0.05,0.494,0.05  
  livingroom:
    enabled: true
    ffmpeg:
      inputs:
        - path: rtsp://localhost:8554/livingroom
          input_args: preset-rtsp-restream
          roles:
            - record
            - audio
        - path: rtsp://localhost:8554/livingroom_sub
          input_args: preset-rtsp-restream
          roles:
            - detect
    record:
      enabled: true
      retain:
        days: 2
        mode: motion
    detect:
      enabled: true
      width: 640
      height: 480
      fps: 4
    objects:
      track:
        - person
        - dog
      filters:
        person:
          min_score: 0.4
          threshold: 0.7
          min_area: 700
```

**Comment by dominikgrygiel** [comment]
> > > Has anyone figured out a way to always record the sub stream and use the main stream only for events? Or do the devs have any plans to implement this?
> > 
> > 
> > I was able to achieve it by duplicating the camera:
> > cameras:
> > camera_one_continuous:
> > enabled: true
> > ffmpeg:
> > inputs:
> > - path:
> > rtsp://secondary_stream_url
> > roles:
> > - record
> > detect:
> > enabled: false
> > record:
> > enabled: true
> > retain:
> > days: 7
> > mode: all
> > camera_one:
> > enabled: true
> > ffmpeg:
> > inputs:
> > - path:
> > rtsp://main_stream_url
> > roles:
> > - record
> > - path:
> > rtsp://secondary_stream_url
> > roles:
> > - detect
> > detect:
> > enabled: true
> > record:
> > enabled: true
> > retain:
> > days: 30
> > mode: motion
> 
> I tried to replicate this on version 0.16 and it appears that the main stream is still continuing to record even though it is set to motion. When I go to the Live / History, and scrub the timeline, i am able to play the video on the main stream. The only difference I could tell between the continues sub stream and continues (supposed to be just motion) main stream, the main stream has a lot of motion yellow horizontal lines in its timeline. Here is my settings:
> 
> ```
> cameras:
>   livingroom_sub_continues:
>     enabled: true
>     ui:
>       dashboard: false
>     ffmpeg:
>       inputs:
>         - path: rtsp://localhost:8554/livingroom_sub
>           input_args: preset-rtsp-restream
>           roles:
>             - record
>             - audio
>     record:
>       enabled: true
>       retain:
>         days: 2
>         mode: all
>     detect:
>       enabled: true
>       width: 640
>       height: 480
>       fps: 4
>     objects:
>       track:
>         - person
>         - dog
>       filters:
>         person:
>           min_score: 0.4
>           threshold: 0.7
>           min_area: 700
>     motion:
>       mask: 0.493,0.019,0.594,0.017,0.594,0.05,0.494,0.05  
>   livingroom:
>     enabled: true
>     ffmpeg:
>       inputs:
>         - path: rtsp://localhost:8554/livingroom
>           input_args: preset-rtsp-restream
>           roles:
>             - record
>             - audio
>         - path: rtsp://localhost:8554/livingroom_sub
>           input_args: preset-rtsp-restream
>           roles:
>             - detect
>     record:
>       enabled: true
>       retain:
>         days: 2
>         mode: motion
>     detect:
>       enabled: true
>       width: 640
>       height: 480
>       fps: 4
>     objects:
>       track:
>         - person
>         - dog
>       filters:
>         person:
>           min_score: 0.4
>           threshold: 0.7
>           min_area: 700
> ```

@kabaga I'm running on 0.16.2. The main difference between our configs is that you have the `audio` role, which I don't have and not sure how this impacts recording - maybe something to experiment with?

Also, when I go to the Live / History I can "jog" through the main stream recordings and I do see some frames from the secondary stream there. However, when I start playing, it plays the main stream and jumps through the history when it detected motion. Not sure how Frigate does that and if it detects that there are already recordings from the other stream it can use for preview.

I've checked the actual recording files on the disk and I can confirm that the main stream only gets recorded when motion is detected (ex. no recordings during the night).

Remember, that the way we have it configured now, it should be treated as 2 separate cameras, so you could try commenting out/removing the `livingroom_sub_continues`, so that you can focus on fixing the recording behaviour on the main stream. Going to "Settings" -> "Debug" might be helpful. Maybe ex. lighting change (clouds covering the sun) could be triggering motion and recording?

**Comment by marrobHD** [comment]
I'd like to see this coming. Some cameras now feature a third stream for a second camera which has analog zoom applied. I'd like to have the main and special stream recorded.

**Comment by 3ricj** [comment]
FYI - I built this substream system out completely and tested it extensively -- but because the developers (who haven't been doing anything on this feature in 5 years)  - didn't get consulted first they rejected my contributions outright - didn't make recommendations for what I should change, or even provided a single byte of meaningful feedback other than I didn't check with them first before working on something they have abandoned.  You can see this here:  https://github.com/blakeblackshear/frigate/pull/23046

**Comment by NickM-27** [comment]
That is not an accurate representation of what happened. The contributing documentation specifically states that a discussion should be created before putting up a PR for a large feature, this is also reiterated in the PR template. This is both so that we can align on the implementation details as well as to save the contributor time and not write / generate code that does not align with our existing plans / the existing Frigate architecture.

Like was said in the comments, we already have experimental implementations of this running locally that we are testing on our hardware and considering carefully. The multi-stream playback feature is not planned to be implemented by us until a future version of Frigate as we currently have other things we are prioritizing.  0.18 is already feature frozen and brings many highly requested features from users as well as full UI configuration which is the number one feedback we get.

I explained multiple technical issues with your implementation and suggested creating a discussion, it is much easier to build up from a plan to an implementation than to break down an implementation and figure out everything that needs to change.

We did our best to be polite and explain our perspective, we (the maintainers of Frigate) are ultimately responsible for what gets merged and supported long-term. For that reason, we are selective about changes and prioritize those that align with the project’s direction and can be responsibly maintained by us in the long term.

**Comment by 3ricj** [comment]
> That is not an accurate representation of what happened. The contributing documentation specifically states that a discussion should be created before putting up a PR for a large feature, this is also reiterated in the PR template. This is both so that we can align on the implementation details as well as to save the contributor time and not write / generate code that does not align with our existing plans / the existing Frigate architecture.


While, true, also not.  The feature requests on github have years of comments from end users trying to make it work, and zero comments from developers about working on this.  There was no discussion happening anywhere.  It was clearly an abandoned feature request.  Or perhaps it's just happening in secret -- but that's not much of a difference. 

> Like was said in the comments, we already have experimental implementations of this running locally that we are testing on our hardware and considering carefully. The multi-stream playback feature is not planned to be implemented by us until a future version of Frigate as we currently have other things we are prioritizing. 0.18 is already feature frozen and brings many highly requested features from users as well as full UI configuration which is the number one feedback we get.

Well, you must have been working on this in secret.  There are no code checkins, forks, PRs, comments, or any public information about this anywhere.   If I'm missing something, please provide a link. 

> I explained multiple technical issues with your implementation and suggested creating a discussion, it is much easier to build up from a plan to an implementation than to break down an implementation and figure out everything that needs to change.

No, you did not raise a signal technical issue.  I raised them in my comments, and got zero helpful replies. 

> We did our best to be polite and explain our perspective, we (the maintainers of Frigate) are ultimately responsible for what gets merged and supported long-term. For that reason, we are selective about changes and prioritize those that align with the project’s direction and can be responsibly maintained by us in the long term.

Yes, we agree here - but you had a few options when faced with something like this: 

1.  You could accept it and merge it.   You didn't want to do that, no problem. 
2.  You could provide clear feedback for what needs to change about it for it to be helpful.  You decided... it would take too much time to give feedback? 
3.  You could say as built, it doesn't meet our needs for maintainability, what we could use is something like this ->  Point to clear documentation of what's needed or what needs to be tested or built or thought about etc.  "we are working on it in secret" isn't that. 
4.  Just reject the PR outright waving your hands that you didn't get consulted first, how dare I try to help this project without consulting the mighty maintainers first?  

In short -- I don't fault you for not accepting my PR -- that's just fine, and not even remotely unexpected.  What was unexpected was how you handled it going forward.  I am motivated to make a system like this work -- but you effectively ejected me from the "secret club" who is working on this.  That's your right -- but it doesn't harness the energy and time offered.

**Comment by thorleifjacobsen** [comment]
Is this feature beeing worked on? Really eating my disk. I had 30 days retention on BlueIris and testing this and not able to do sub -> main swap on motion makes me want to go back to BlueIris.

**Comment by NickM-27** [comment]
0.18 is feature frozen, so at a minimum it won't be implemented until 0.19. We have two releases per year.

**Comment by thorleifjacobsen** [comment]
> 0.18 is feature frozen, so at a minimum it won't be implemented until 0.19. We have two releases per year. 

This is 5 year old request. Meaning; 10 releases old? And still not implemented?

**Comment by NickM-27** [comment]
This is free and open software, which was started and worked on for years by primarily a single person, and more recently is worked on primarily by two people in our free time.

This feature in particular has a large surface area and touches many different parts of the application, along with many requirements for it to work well in a variety of use cases and situations. In past releases the main focus was on stability and improving the experience of existing features. That focus has now shifted to being on making configuration easier, with full UI config coming in 0.18. This is another item that is often question as "how is this still not implemented?".

The simple fact is we work on Frigate in our free time and are in no rush to get sloppy features out the door just to mark something as done. This feature is planned, but it is not the current priority.

**Comment by 3ricj** [comment]
> > 0.18 is feature frozen, so at a minimum it won't be implemented until 0.19. We have two releases per year.
> 
> This is 5 year old request. Meaning; 10 releases old? And still not implemented?

Yeah - it's a major problem - it's why I spent time building out this feature - in the end, the developers rejected my open source contribution that fully implemented this because I didn't ask permission from them first to build something.  It's available as a patch if you'd like to play with it and give me feedback.

**Comment by raidflex** [comment]
Considering how much storage costs right now and the huge waste of space 4k video takes up is there a higher priority now with this feature? This seems like a core function of a CCTV system.

Considering just a couple 4k cameras can eat up TBs of storage for just a couple days, it's a massive waste of space. But you still need to be able to identify an object or person so the higher the quality of the alert the better.

I understand this is a free product, but 5 years is a long time for a feature request and for something that has a direct affect on the cost of the hardware required.

As well as Blue Iris worked, I hate that I needed to use the microslop that is Windows just to run it. 

I don't mind paying for a product either.

**Comment by 3ricj** [comment]
> Considering how much storage costs right now and the huge waste of space 4k video takes up is there a higher priority now with this feature? This seems like a core function of a CCTV system.
> 
> Considering just a couple 4k cameras can eat up TBs of storage for just a couple days, it's a massive waste of space. But you still need to be able to identify an object or person so the higher the quality of the alert the better.
> 
> I understand this is a free product, but 5 years is a long time for a feature request and for something that has a direct affect on the cost of the hardware required.
> 
> As well as Blue Iris worked, I hate that I needed to use the microslop that is Windows just to run it.
> 
> I don't mind paying for a product either.

They will not be building this feature

**Comment by raidflex** [comment]
> > Considering how much storage costs right now and the huge waste of space 4k video takes up is there a higher priority now with this feature? This seems like a core function of a CCTV system.
> > Considering just a couple 4k cameras can eat up TBs of storage for just a couple days, it's a massive waste of space. But you still need to be able to identify an object or person so the higher the quality of the alert the better.
> > I understand this is a free product, but 5 years is a long time for a feature request and for something that has a direct affect on the cost of the hardware required.
> > As well as Blue Iris worked, I hate that I needed to use the microslop that is Windows just to run it.
> > I don't mind paying for a product either.
> 
> They will not be building this feature

I'm curious on the reasoning behind this, if that is the case. If the point of Frigate is to have a fully local CCTV system then storage space is at the core of that and direct cost of using the product. Times have also changed and I doubt storage costs will be going down anytime soon, if ever.

Improving the usability and UI is important don't get me wrong, but also if you cannot afford the storage to make it work in a reliable way then what's the point.

**Comment by hawkeye217** [comment]
> They will not be building this feature

That is incorrect. This feature request is still open and marked as planned, which indicates that we have intention to build it out in some form for a future version of Frigate.

**Comment by raidflex** [comment]
> > They will not be building this feature
> 
> That is incorrect. This feature request is still open and marked as planned, which indicates that we have intention to build it out in some form for a future version of Frigate.

Well that is good hear, but considering it's been 5 years I hope it will happen in the not so distant future. I also understand that rushing and creating buggy software is not good either.

Storage cost now is a large barrier of entry with a a local system, which is unfortunately the climate we are in.

**Comment by darkalfx** [comment]
> > They will not be building this feature
> 
> That is incorrect. This feature request is still open and marked as planned, which indicates that we have intention to build it out in some form for a future version of Frigate.

When??

**Comment by hawkeye217** [comment]
There is no time frame available, but again, a feature being marked as planned means it is coming. We have a number of improvements in the works that overlap in this space, each of which needs careful planning and testing.

**Comment by 3ricj** [comment]
> There is no time frame available, but again, a feature being marked as planned means it is coming. We have a number of improvements in the works that overlap in this space, each of which needs careful planning and testing.

I already built it for you. Stop lying to your customers.

**Comment by 3ricj** [comment]
Why are you deleting my comments?

**Comment by hawkeye217** [comment]
Your comments above were hidden as outdated because we will in fact be building this feature, and your comments indicated that we will not be.

**Comment by partytimeexcellent** [comment]
Can’t wait for this! Each new release I eagerly jump to the change log, only to have my hopes dashed again.

**Comment by trullock** [comment]
Thanks for your hard work on Frigate

+1 for this feature

**Comment by trullock** [comment]
Woo!

**Comment by mjleonides** [comment]
Very exciting, can’t wait to give it a shot and get off Windows forever.

**Comment by raidflex** [comment]
Nice!

**Comment by GoodKnight4** [comment]
Thank you! I have been eagerly waiting for this feature!
